### ISSUE TITLE
Support configuration of convolution dimension numbers

### ISSUE THREAD

**Issue by seanmor5** [elixir-nx/nx#262]
Right now, our convolutions require input and kernel to have shapes:

- `input_shape - (batch_size, input_channels, spatial_dim0, ..., spatial_dimN)`
- `kernel_shape - (output_channels, input_channels, kernel_spatial_dim0, ..., kernel_spatial_dimN)`

With a resulting output shape of:

- `output_shape - (batch_size, output_channels, output_spatial_dim0, ..., output_spatial_dimN)`

While this is probably a pretty good general configuration, some backends or compilers may enforce other shape constraints. It would be best to support a configuration option that allows users to choose how to configure their convolutions. As an example, here's a JAX convolution dimension number configuration that matches the semantics described above:

```python
dimnos = ('NCHW', 'OIHW', 'NCHW')
```

There are two options for supporting this configuration. The first is to add an option to `Nx.conv` which interprets the configuration to compute shape constraints, padding configuration, output shape, etc. and then passes the same configuration to the backend or compiler to handle. In the case of an unsupported configuration, the backend/compiler will simply need to raise.

The other option is to have backends/compilers configure a default convolution dimension configuration and always dispatch to that implementation. The advantage here, I think, is portability of numerical definitions. Rather than needing to change inputs to `Nx.conv` based on a compiler/backend, you'd only need to set your backends default convolution configuration.

Implementation note for both EXLA and Nx: EXLA supports general configurations should implementation should be straightforward. For the Binary backend, I'm in favor of raising on shapes that don't match our current configuration, mainly to save the overhead of having to refactor and rewrite a convolution from scratch.

**Comment by josevalim** [comment]
For the binary backend, could we transpose after the fact if you provide a different configuration? Or does this ultimately affect the computed result?

**Comment by seanmor5** [comment]
Good point, we could definitely transpose before/after the convolution to support different configurations.

**Comment by josevalim** [comment]
Btw, related to the backend API, we should probably standardize on something similar to what XLA exposes.

**Comment by seanmor5** [comment]
@josevalim A simple proposal to address this is to expose permutations as an input option rather than dealing with a potentially confusing configuration with tuple/list of tuples/lists. We can normalize the permutations in the same way as in `transpose`, wdyt?

Tackling this issue will make implementing the gradient for convolution a lot easier

**Comment by josevalim** [comment]
@seanmor5 sounds good to me!

**Comment by elbow-jason** [comment]
Notes for batch (N), channels (C), and spatial dims (HW) support in some different frameworks.

[torch - shape section](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html)

  + input:  {N, C, HW}
  + output: {N, C, HW}

[tensorflow - data_format](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d)


  + input: {N, HW, C}
  + output: {N, HW, C}

and:

  + input: {N, C, HW}
  + output: {N, C, HW}

[Jax - dimension_numbers](https://github.com/google/jax/blob/8ad511822808138a0f07ead29db5a8403189005b/jax/_src/lax/lax.py#L559-L578)

  + input:  {N, C, HW}
  + kernel: {O, I, HW}
  + output: {N, C, HW}

and:
  + input:  {N, HW, C}
  + kernel: {HW, I, O}
  + output: {N, HW, C}
